### PR TITLE
Exclude plugin vendor directories from openapi search paths

### PR DESCRIPTION
**Pull request by Vinai**
### 1. Why is this change necessary?

When installing dependencies in plugins in order to execute unit tests in isolation from the shopware development system, the OpenApiLoader used up all available memory (tested with 16G memory_limit).

I agree this is an edge case, but it cost me 4-5 hours to figure out why the swagger UI wasn't loading. I hope this will save someone else the trouble.

### 2. What does this change do, exactly?

This commit excludes vendor directories within custom/plugins/* from the OpenApiLoader search path.

### 3. Describe each step to reproduce the issue or behaviour.

1. Create a plugin that requires `"shopware/platform": "6.1.x@dev"` in its `composer.json.
2. Run `composer install` in the plugin directory
3. Try to open the swagger UI at /api/v1/_info/swagger.html

### 4. Please link to the relevant issues (if any).

No issue.

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes <-- none are required
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by Vinai** [comment]
Note: this PR is partially also motivated by me wanting to learn about the shopware contribution process.
Please let me know what I can improve.

**Comment by Vinai** [comment]
As far as I can tell the failing tests are not related to this PR.

**Comment by pweyck** [comment]
Hello @Vinai, 

thanks for the PR and particularly for the test :)

Our gitlab pipeline is green except for some styling issues. Please use our pre-commit hook or run cs-fixer like this: https://github.com/shopware/development/blob/79b602871d41cfdfb6cb073e09f2599aaad6d99e/dev-ops/pre-commit#L60

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+ * It allows using the test Flysystem instance with file system related code expecting string paths.
+ * This class only implements the methods for reading files and directories, writes have to be
+ * done using the Flysystem instance directly.
+ */
+class FlysystemReadOnlyStreamWrapper
+{
+    use KernelTestBehaviour;
+    use FilesystemBehaviour;
+
+    const MODES = [
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: please add a visibility

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        'file' => 0100644, // rw-r--r--
+        'dir' => 0040755,  // rwxr-xr-x
+    ];
+
+    /**
+     * @var string
+     */
+    private $filePath;
+
+    private $fileOffset = 0;
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php:35]: add var annotations to all properties

**Comment by mitelg** [comment]
hey @Vinai,
I will also leave a few comments in a while 😉✌️

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+use Shopware\Core\Framework\Test\TestCaseBase\KernelTestBehaviour;
+
+/**
+ * Incomplete adapter to make Flysystem usable via stream wrapper.
+ *
+ * It allows using the test Flysystem instance with file system related code expecting string paths.
+ * This class only implements the methods for reading files and directories, writes have to be
+ * done using the Flysystem instance directly.
+ */
+class FlysystemReadOnlyStreamWrapper
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: where is this class used?

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+    /**
+     * @throws FileNotFoundException
+     */
+    private function getFileContents(): string
+    {
+        return $this->getFilesystemInstance()->read($this->filePath);
+    }
+
+    public function stream_open(string $path, string $mode, int $options, &$opened_path): bool
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: please use camcelCase method names

Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: parameter should also have camelCase names

add a typehint for the last parameter

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        return $buffer;
+    }
+
+    public function stream_tell(): int
+    {
+        return $this->fileOffset;
+    }
+
+    /**
+     * @throws FileNotFoundException
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: this method never throws this exception

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        }
+    }
+
+    private function getSize(): int
+    {
+        return $this->getMetadata($this->filePath)['size'];
+    }
+
+    /**
+     * @throws FileNotFoundException
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: exception is not thrown in this method

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        }
+
+        $this->filePath   = $filePath;
+        $this->fileOffset = 0;
+
+        return true;
+    }
+
+    /**
+     * @throws FileNotFoundException
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: we only annotate exceptions if they are thrown in this method

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/OpenApiLoaderTest.php [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+use PHPUnit\Framework\TestCase;
+use Shopware\Core\Framework\Api\ApiDefinition\Generator\OpenApi\OpenApiLoader;
+use Shopware\Core\Framework\Api\Controller\ApiController;
+use Shopware\Core\Framework\Test\TestCaseBase\IntegrationTestBehaviour;
+
+class OpenApiLoaderTest extends TestCase
+{
+    use IntegrationTestBehaviour;
+
+    const TESTFS = 'testfs';
```
Comment (mitelg) [inline src/Core/Framework/Test/Api/ApiDefinition/Generator/OpenApiLoaderTest.php]: visibility

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php (hunk 1 above)
Comment (Vinai) [reply src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: done

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php (hunk 2 above)
Comment (Vinai) [reply src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php:35]: done

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php (hunk 4 above)
Comment (Vinai) [reply src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: This is not possible since the function names are defined by the PHP [stream wrapper API](https://www.php.net/manual/en/class.streamwrapper.php).

Comment (Vinai) [reply src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: done

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php (hunk 7 above)
Comment (Vinai) [reply src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: PHPStorm doesn't like that, but done :)

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/OpenApiLoaderTest.php (hunk 8 above)
Comment (Vinai) [reply src/Core/Framework/Test/Api/ApiDefinition/Generator/OpenApiLoaderTest.php]: done

**Comment by Vinai** [comment]
In gitter @shyim just told me about a [patch he made](https://gist.github.com/shyim/39d491ea527d457ab67d4e6e6adb495c) for the issue which, in addition to resolving the problem, also reduces the load time a lot by reducing the number of files to scan to only active controllers. His is probably the better solution.

File: src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php (hunk 4 above)
Comment (mitelg) [reply src/Core/Framework/Test/Api/ApiDefinition/Generator/FlysystemReadOnlyStreamWrapper.php]: oh, ok :smile:

**Comment by mitelg** [comment]
ok, maybe we could use your test with the helper class nevertheless

@shyim @pweyck what do you think?

**Comment by Vinai** [comment]
@mitelg I think the test would pass with the patch, however, it wouldn't test the actually behavior of the patch. For example, the API controller that is copied into the in-memory file system isn't used to add the `_search` path to the result, but instead that would come from the registered controller... so that part of the fixture could be removed without influencing the test.
The only thing of value this PR might still contain is the flysystem wrapper, but since it's not used, it might as well be kept out of the system.
So all in all I think it's better if I close the PR. Thank you for the reviews, I learned something and it was a good experience.
